migrations/20230412200134_feature_grants__create.sql

```diff
@@ -0,0 +1,45 @@
+CREATE TYPE feature_grant_type AS ENUM('automatic', 'manual', 'subscription');
+
+CREATE TABLE IF NOT EXISTS
+  feature_grants (
+    id BIGSERIAL PRIMARY KEY,
+    name TEXT NOT NULL,
+    type feature_grant_type NOT NULL,
+    created_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(),
+    institution_id BIGINT REFERENCES institutions ON UPDATE CASCADE ON DELETE CASCADE,
+    course_id BIGINT REFERENCES pl_courses ON UPDATE CASCADE ON DELETE CASCADE,
+    course_instance_id BIGINT REFERENCES course_instances ON UPDATE CASCADE ON DELETE CASCADE,
+    user_id BIGINT REFERENCES users ON UPDATE CASCADE ON DELETE CASCADE,
+    CONSTRAINT feature_grants_by_name_idx UNIQUE (
+      name,
+      institution_id,
+      course_id,
+      course_instance_id,
+      user_id,
+      type
+    )
+  );
+
+-- Secondary index to allow looking up all feature flags for a given entity.
+CREATE INDEX IF NOT EXISTS feature_grants_by_entity_idx ON feature_grants (
+  institution_id,
+  course_id,
+  course_instance_id,
+  user_id,
+  name,
+  type
+);
+
+-- Secondary index to allow looking up all feature flags for a given user.
+CREATE INDEX IF NOT EXISTS feature_grants_user_id_name_type_idx ON feature_grants (
+  user_id,
+  name,
+  type
+);
+
+-- Secondary index to allow looking up all users with a given feature flag.
+CREATE INDEX IF NOT EXISTS feature_grants_name_user_id_type_idx ON feature_grants (
+  name,
+  user_id,
+  type
+);
```

Comment:
These probably aren't strictly necessary at this point, I'm just thinking ahead to various query patterns we'll want to have in the near future. I imagine this table could get quite large, though maybe not for a while. Some thoughts/comments:

- `type` is at the end because we might frequently want to filter out specific types (e.g. we might not want to show a list of all `subscription` feature grants on an admin page by default).
- We might be able to avoid the two user-specific indexes if we put `user_id` first in the other one?

Reply:
One way around this would be to replace `institution_id`/`course_id`/`course_instance_id` with a single `key` column that has values like `course:1234` or `course_instance:57`. This is more flexible at the cost of losing referential integrity 😢

I'm in favor of keeping things tied to our particular hierarchy at the moment (and maintaining RI!).